w08h03/test/pgdp/datastructures/RandomTest.java

```diff
@@ -0,0 +1,109 @@
+package pgdp.datastructures;
+
+import org.junit.jupiter.api.RepeatedTest;
+
+import java.nio.charset.StandardCharsets;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.Random;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class RandomTest {
+    @RepeatedTest(10000) // amount tests
+    public void randomlyTestIntValues() {
+        ArrayList<Integer> values = new ArrayList<>();
+        ArrayList<Integer> solution = new ArrayList<>();
+
+        Random num = new Random();

```

Comment:
Bitte irgend ein Seed festlegen

Reply:
Warum? Der Test produziert keine Outputs die man vergleichen könnte und auch allgemein würde immer derselbe Seed doch dem wahllosen Testen von Bäumen schaden - dann würde einfach nur 10 000 mal derselbe Baum erstellt werden.

Reply:
Sehe ich eigentlich wie @blubsbuar, aber es würde auch nicht wehtun.

Reply:
Wenn `Random` in Tests verwenden und den Seed nicht setzen, ist der Test nicht mehr deterministisch. Es könnte passieren, dass eine bestimmte Zahlenfolge nicht funktioniert, wodurch der Test failt. Man ändert an seiner Implementierung gar nichts, führt den Test wieder aus und auf einmal funktioniert der Test. Die Wahrscheinlichkeit bei 10.000 ist zwar gering, aber kann dennoch passieren.

Das Setzen eines Seeds macht den Test wieder deterministisch.

Reply:
[Some good arguments for using a seed which I won't bother to summarize here](https://softwareengineering.stackexchange.com/questions/429601/is-the-usage-of-random-values-in-unit-testing-a-good-practice)

Reply:
> Wenn wir `Random` in Tests verwenden und den Seed nicht setzen, ist der Test nicht mehr deterministisch. Es könnte passieren, dass eine bestimmte Zahlenfolge nicht funktioniert, wodurch der Test failt. Man ändert an seiner Implementierung gar nichts, führt den Test wieder aus und auf einmal funktioniert der Test. Die Wahrscheinlichkeit bei 10.000 ist zwar gering, aber kann dennoch passieren.
> 
> Das Setzen eines Seeds macht den Test wieder deterministisch.

Wenn der Test eine Zahlenfolge findet, für die dein Code nicht funktioniert, ist das doch im Grunde genommen genau das, worauf diese brute force Spielerei abzielt. Würde man hier einen fixen Seed festlegen, könnte man sich die 39996 übrigen Testausführungen sparen, weil immer dieselben Werte generiert werden würden. Die Tests berechnen die Lösung aber dynamisch (natürlich ohne, dass die Aufgabenstellung gelöst wird), weshalb sie eigentlich trotzdem noch korrekt sein müssen. Oder übersehe ich hier etwas?

Reply:
> [Some good arguments for using a seed which I won't bother to summarize here](https://softwareengineering.stackexchange.com/questions/429601/is-the-usage-of-random-values-in-unit-testing-a-good-practice)

Auszug aus dem Link: "Fuzz testing, and therefore Autofixture, has a separate value: bug discovery. Here, randomization is actually desirable because it can help you find edge cases that you hadn't even anticipated." - Das ist doch genau, was ich mit diesen Tests versuche zu erreichen. Ein Seed würde einfach keinen Sinn machen.

Reply:
>  Würde man hier einen fixen Seed festlegen, könnte man sich die 39996 übrigen Testausführungen sparen, weil immer dieselben Werte generiert werden würden.

Dafür ist ein Seed nicht da. 

Das feste Setzen eines Seed ist dafür da, dass man bestimmt, dass die Generierung von zufälligen Zahlen immer nach derselben Abfolge passiert. Sprich, wenn ich `Random(100).next(5)` aufrufe (Seed = 100), bekomme ich eine zufällige Zahl zwischen 0 und 4, z.B. 2. Dadurch ich denn Seed auf 100 festgelegt habe, weiß ich, dass ich wieder 2 bekomme, wenn ich `Random(100).next(5)` (ich erstelle ein _neues_ `Random` Objekt) aufrufe. Jedoch heißt das nicht, dass ich wieder 2 bekomme, ich mit _demselben_ `Random` Objekt `nextInt` aufrufe.

```java
import java.util.Random;

public class MyClass {
    public static void main(String args[]) {

      // Es ist wichtig, dass wir das Random Objekt nicht in der for-loop 
      // erstellen, sondern außerhalb
      Random r = new Random(100);

      for(int i = 0; i < 10; i++) {
          int randomInt = r.nextInt(1000);
          System.out.println(randomInt);
      }
    }
}
``` 

Du kannst den Code hier selber online ausprobieren: https://www.jdoodle.com/iembed/v0/BiO

Diese for-loop wird _immer_ als Zahlenfolge folgende Zahlen ausgeben: "915, 250, 874, 988, 291, 666, 36, 288, 723, 713". Jedoch ist jede Zahl an sich zufällig gewählt worden (es ist nicht im 915, 915, 915, 915), weil nur 1 `Ranom` Objekt erstellt haben. Würden wir *in* der for loop jedes mal ein neues `Random` Objekt erstellen, dann wäre es so wie du sagtest, dass jedes mal die Zahl 915 rauskommen würde.

Reply:
Ja, das ist mir schon klar, aber würde ich einen Seed festlegen, dann würden meine Tests 10000 mal (wegen RepeatedTest(10000), was die Methode bzw. den Test 10000 mal aufruft -> also auch das Objekt 10000mal erstellt) 10000mal ein Random Objekt mit demselben Seed erstellen, wodurch alle 10000 Tests dieselben Zahlen testen  wie der 1.

Reply:
Wenn das Erstellen des `Random` Objektes aus dem Test ziehst, solltest du für jeden Test-Aufruf andere Wert bekommen, denn noch deterministisch die Werte erstellen können:

```java
Random randomTestIntValues = new Random();

@RepeatedTest(10000) // amount tests
public void randomlyTestIntValues() {
    ArrayList<Integer> values = new ArrayList<>();
    ArrayList<Integer> solution = new ArrayList<>();

    int amountOfValues = randomTestIntValues.nextInt(250); // max amount of values the tree should contain
    for (int i = 0; i < amountOfValues; i++) {
        values.add(randomTestIntValues.nextInt());
        solution = new ArrayList<>(values);
        Collections.sort(solution);
    }
    QuarternarySearchTree<Integer> n = new QuarternarySearchTree<Integer>();
    for (Integer i : values) {
        n.insert(i);
    }
    ArrayList<Integer> actual = new ArrayList<>();
    for (Integer r : n) {
        actual.add(r);
    }
    assertEquals(actual, solution);
}
```

Du müsstest nur darauf achten, dass du _nicht_ ein `Random` für einen weiteren Test (z.B. `randomlyTestDoubleValues()`) weiterverwendet, weil dies zu anderen Werten führen würde, wenn man die Tests in einer anderen Reihenfolge ausführt.

Reply:
Gut, da hast du einen Punkt, werde es dahingehend anpassen

Reply:
> > [Some good arguments for using a seed which I won't bother to summarize here](https://softwareengineering.stackexchange.com/questions/429601/is-the-usage-of-random-values-in-unit-testing-a-good-practice)
> 
> Auszug aus dem Link: "Fuzz testing, and therefore Autofixture, has a separate value: bug discovery. Here, randomization is actually desirable because it can help you find edge cases that you hadn't even anticipated." - Das ist doch genau, was ich mit diesen Tests versuche zu erreichen. Ein Seed würde einfach keinen Sinn machen.

Ich spreche mich ja auch nicht gegen randomisierte werte aus... Durch einen seed werden diese ja auch nicht weniger zufällig. Lediglich reproduzierbar zufällig, also das beste aus beiden Welten. Das Zitat stellt also kein Argument gegen einen festgelegten Seed dar.